x-pack/solutions/observability/plugins/apm/common/service_map/get_service_map_nodes.ts

```diff
@@ -0,0 +1,292 @@
+import {
+  SERVICE_NAME,
+  SPAN_DESTINATION_SERVICE_RESOURCE,
+  SPAN_TYPE,
+  SPAN_SUBTYPE,
+} from '../es_fields/apm';
+import type { ExitSpanDestination, ServicesResponse } from './types';
+import type {
+  Connection,
+  ConnectionNode,
+  ServiceConnectionNode,
+  ExternalConnectionNode,
+  ConnectionElement,
+  ConnectionEdge,
+  ServiceMapConnections,
+  GroupResourceNodesResponse,
+} from './types';
+
+import { groupResourceNodes } from './group_resource_nodes';
+import { getConnectionId, isExitSpan } from './utils';
+
+const FORBIDDEN_SERVICE_NAMES = ['constructor'];
+
+function addMessagingConnections(
+  connections: Connection[],
+  destinationServices: ExitSpanDestination[]
+): Connection[] {
+  const serviceMap = new Map(
+    destinationServices.map(({ from, to }) => [from[SPAN_DESTINATION_SERVICE_RESOURCE], to])
+  );
+
+  const messagingConnections = connections.reduce<Connection[]>((acc, connection) => {
+    const destination = connection.destination;
+    if (isExitSpan(destination) && destination[SPAN_TYPE] === 'messaging') {
+      const matchedService = serviceMap.get(destination[SPAN_DESTINATION_SERVICE_RESOURCE]);
+      if (matchedService) {
+        acc.push({
+          source: destination,
+          destination: matchedService,
+        });
+      }
+    }
+    return acc;
+  }, []);
+
+  return [...connections, ...messagingConnections];
+}
+
+function getAllNodes(services: ServicesResponse[], connections: Connection[]) {
+  const allNodesMap = new Map<string, ConnectionNode>();
+
+  connections.forEach((connection) => {
+    const sourceId = connection.source.id;
+    const destinationId = connection.destination.id;
+    if (!allNodesMap.has(sourceId)) {
+      allNodesMap.set(sourceId, { ...connection.source, id: sourceId });
+    }
+    if (!allNodesMap.has(destinationId)) {
+      allNodesMap.set(destinationId, { ...connection.destination, id: destinationId });
+    }
+  });
+
+  // Derive the rest of the map nodes from the connections and add the services
+  // from the services data query
+  services
+    .filter((service) => !FORBIDDEN_SERVICE_NAMES.includes(service[SERVICE_NAME]))
+    .forEach((service) => {
+      const id = service[SERVICE_NAME];
+      if (!allNodesMap.has(id)) {
+        allNodesMap.set(id, { ...service, id });
+      }
+    });
+
+  return allNodesMap;
+}
+
+function getAllServices(
+  allNodes: Map<string, ConnectionNode>,
+  destinationServices: ExitSpanDestination[]
+) {
+  const connectionFromDiscoveredServices: ServiceConnectionNode[] = [];
+
+  for (const { from, to } of destinationServices) {
+    const fromId = from.id;
+    const toId = to.id;
+
+    if (allNodes.has(fromId) && !allNodes.has(toId)) {
+      connectionFromDiscoveredServices.push({ ...to, id: toId });
+    }
+  }
+
+  // List of nodes that are services
+  const serviceNodes = new Map(
+    [...Array.from(allNodes.values()), ...connectionFromDiscoveredServices]
```

Comment:
`.values()` returns an iterator, so with spreading into an array, `Array.from` is redundant and adds an extra allocation that can be avoided.
```suggestion
    [...allNodes.values(), ...connectionFromDiscoveredServices]
```